### PR TITLE
Add wallet network settings;  overhaul electrum

### PR DESCRIPTION
**Pull request by LLFourn**
- **settings page** available everyhere
- **multiple networks** we support all testnets 
- **Electrum** connections more robust because we model the state of the connection, reconnecting if need be. Also has a status icon in the bar.
- **mainnet support** settings the wallet's network during keygen

### CODE REVIEW

File: frostsnapp/native/src/api.rs [hunk 1]
```diff
@@ -614,47 +676,81 @@ impl Wallet {
...
-fn _load(
-    db_file: String,
-    usb_serial_manager: UsbSerialManager,
-) -> Result<(Coordinator, Wallet, BitcoinContext)> {
-    event!(Level::INFO, path = db_file, "initializing database");
+#[derive(Debug, Clone)]
+pub struct BitcoinNetwork(pub RustOpaque<RBitcoinNetwork>);
 
+/// The point of this is to keep bitcoin API functionalities that don't require the wallet separate
+/// from it.
```
Comment (nickfarrow) [inline frostsnapp/native/src/api.rs:698]: Nice, good idea

File: frostsnap_coordinator/src/bitcoin/chain_sync.rs [hunk 2]
```diff
@@ -1,69 +1,336 @@
...
+    status_sink: Box<dyn Sink<ChainStatus>>,
+}
 
-    pub fn broadcast(&self, tx: &bitcoin::Transaction) -> Result<()> {
+enum ConnectionState {
+    Disconnected,
+    Connected(Box<BdkElectrumClient<electrum_client::Client>>),
+}
+
+/// This had to be a macro so that we can precisely mutably borrow certain fields.
```
Comment (nickfarrow) [inline frostsnap_coordinator/src/bitcoin/chain_sync.rs:123]: Cool trick!

File: frostsnap_coordinator/src/frostsnap_persist.rs [hunk 3]
```diff
@@ -28,15 +28,6 @@ impl Persist<rusqlite::Connection> for FrostCoordinator {
             [],
         )?;
 
-        conn.execute(
-            "CREATE TABLE IF NOT EXISTS fs_signing_sessions (
```
Comment (nickfarrow) [inline frostsnap_coordinator/src/frostsnap_persist.rs:32]: we were not using this anywhere? We have `fs_signing_session_state`

File: frostsnap_coordinator/src/persist.rs [hunk 4]
```diff
@@ -27,6 +27,21 @@ impl<T> Persisted<T> {
         Ok(ret)
     }
 
+    pub fn mutate2<C, R>(
```
Comment (nickfarrow) [inline frostsnap_coordinator/src/persist.rs:30]: Maybe `mutate_with_update`?

File: frostsnapp/lib/settings.dart [hunk 5]
```diff
@@ -0,0 +1,465 @@
+import 'dart:async';
+
+import 'package:collection/collection.dart';
+import 'package:flutter/cupertino.dart';
```
Comment (nickfarrow) [inline frostsnapp/lib/settings.dart:4]: few unnecessary imports: `cupertino, electrum_server (empty), serialport`

File: frostsnapp/native/src/api.rs [hunk 6]
```diff
@@ -2,21 +2,29 @@ use crate::device_list::DeviceList;
...
+    wallet::FrostsnapWallet,
+};
 pub use frostsnap_coordinator::firmware_upgrade::FirmwareUpgradeConfirmState;
 pub use frostsnap_coordinator::frostsnap_core;
 use frostsnap_coordinator::frostsnap_core::schnorr_fun::fun::hash::HashAdd;
 pub use frostsnap_coordinator::{
-    check_share::CheckShareState, keygen::KeyGenState, signing::SigningState, DeviceChange,
-    PortDesc,
+    check_share::CheckShareState, keygen::KeyGenState, persist::Persisted, signing::SigningState,
+    DeviceChange, PortDesc, Settings as RSettings,
```
Comment (nickfarrow) [inline frostsnapp/native/src/api.rs:27]: Why `RSettings`, `FSCoordinatorSettings`?

File: frostsnapp/native/src/api.rs [hunk 7]
```diff
@@ -564,9 +559,13 @@ impl Wallet {
         };
         let chain_sync = self.chain_sync.clone();
 
-        let update = chain_sync.sync(sync_request)?;
+        let update = chain_sync
+            .sync(sync_request)
+            .inspect_err(|e| event!(Level::ERROR, error = e.to_string(), "syncing error"))?;
```
Comment (nickfarrow) [inline frostsnapp/native/src/api.rs:564]: nice forgot about this function

File: frostsnapp/native/src/chain_sync.rs [hunk 8]
```diff
@@ -1 +0,0 @@
-
```
Comment (nickfarrow) [inline frostsnapp/native/src/chain_sync.rs]: There's a few other empty files:
```
./frostsnapp/lib/error_dialog.dart
./frostsnapp/lib/electrum_server.dart
./frostsnap_coordinator/src/frb_impl.rs
./device/src/graphics/widgets/backup_input_preview.rs
```

**Review by nickfarrow — APPROVED** [review]

File: frostsnapp/native/src/api.rs (hunk 6 above)
Comment (LLFourn) [reply frostsnapp/native/src/api.rs:27]: R stands for Rust. We could make it a convention to prefix it with Fs instead.

File: frostsnap_coordinator/src/frostsnap_persist.rs (hunk 3 above)
Comment (LLFourn) [reply frostsnap_coordinator/src/frostsnap_persist.rs:32]: Yes seemed to have slipped through the cracks.

File: frostsnap_coordinator/src/persist.rs (hunk 4 above)
Comment (LLFourn) [reply frostsnap_coordinator/src/persist.rs:30]: they all have an "update". This is intended to be the one we use. `staged_mutate` and `mutate` should be deleted.

**Pull request merged**